pmlc/target/x86/stencil_tpp_unary.cc

```diff
@@ -209,34 +271,82 @@ class StencilImpl : public pxa::StencilBase {
 
 public:
   explicit StencilImpl(AffineParallelOp op)
-      : StencilBase(
-            op,
-            {
-                pxa::StencilIndexRequirement{
-                    /*idxName=*/"eltwise_i",
-                    /*tilingGenerator=*/pxa::ExactRangeGenerator(),
-                    pxa::IndexStridePredicates{
-                        [](int64_t stride) { return stride > 1; }, // output
-                        [](int64_t stride) {
-                          return (stride == 0 || stride > 1);
-                        }, // input
-                    }},
-                pxa::StencilIndexRequirement{
-                    /*idxName=*/"eltwise_j",
-                    /*tilingGenerator=*/pxa::ExactRangeGenerator(),
-                    pxa::IndexStridePredicates{
-                        [](int64_t stride) { return stride == 1; }, // output
-                        [](int64_t stride) {
-                          return (stride == 1 || stride == 0);
-                        }, // input
-                    }},
-            }) {}
+      : StencilBase(op, {
+                            pxa::StencilIndexRequirement{
+                                /*idxName=*/"eltwise_i",
+                                /*tilingGenerator=*/pxa::ExactRangeGenerator(),
+                                pxa::IndexStridePredicates{
+                                    [](int64_t stride) {
+                                      return (stride == 0 || stride > 1);
+                                    }, // output
+                                    [](int64_t stride) {
+                                      return (stride == 0 || stride > 1);
+                                    }, // input
+                                }},
+                            pxa::StencilIndexRequirement{
+                                /*idxName=*/"eltwise_j",
+                                /*tilingGenerator=*/pxa::ExactRangeGenerator(),
+                                pxa::IndexStridePredicates{
+                                    [](int64_t stride) {
+                                      return (stride == 1 || stride == 0);
+                                    }, // output
+                                    [](int64_t stride) {
+                                      return (stride == 1 || stride == 0);
+                                    }, // input
+                                }},
+                        }) {}
 };
 
 } // namespace
 
+// For 1D stenciling, we still map to 2D TPPs. We can make 2D TPPs do 1D
+// by having the outer dim being 1.
+void addSingleIteration(AffineParallelOp op) {
+  if (op.getIVs().size() == 1) {
```

Comment:
Please prefer an early exit instead of wrapping everything in an 'if'. Can you please add documentation for the method?

Reply:
Thanks for the advice. It is very helpful. Now in resnet50, some operations are not stenciled (e.g, reduce max/add). This PR is focus on "1D" case:
```
%359 = affine.parallel (%arg110) = (0) to (1000) reduce ("assign") -> (memref<1x1xf32>) {
    %360 = pxa.load %354[%c0, %arg110] : memref<1x1000xf32> "softmax"
    %361 = pxa.reduce addf %360, %356[%c0, 0] : memref<1x1xf32> "softmax"
    affine.yield %361 : memref<1x1xf32> "softmax"
} "softmax"
```
As you can see, the loop above has only one inducution variable (%arg110). However, our current StencilImpl requires at least two variables.
```
getOperation().walk([](AffineParallelOp op) {
  if (op.getIVs().size() >= 2) {
    StencilImpl stencil(op);
    stencil.performStenciling();
  }
});
```
My solution is to add a (0, 1) dim (addSingleIteration) so that it can still be mapped to a 2D TPP without major code changes. I will refactor my code and add more comments. Thanks.